apps/middleman/src/actions/Providers.ts

```diff
@@ -19,26 +22,122 @@ const updateProvidersSchema = z.object({
-    const response = await fetch(url);
+    const response = await fetch(providersCdnUrl);
+
     if (!response.ok) {
-      throw new Error(
-        `Failed to fetch providers from CDN: ${response.statusText}`
-      );
+      throw new Error(`Failed to fetch providers: ${response.statusText}`);
+    }
+
+    type CdnProvider = {
+      name: string;
+      identity: string;
+      identityHistory: string[];
+      url: string;
+    };
+
+    const providersFromCdn = (await response.json()) as CdnProvider[];
+    console.log(`[Providers] Fetched ${providersFromCdn.length} providers from CDN`);
+
+    const currentProviders = await list(true);
+
+    const currentProvidersMap = new Map(
+      currentProviders.map((p) => [p.identity, p]),
+    );
+
+    const allCdnIdentities = new Set<string>();
+
+    let inserted = 0;
+    let updated = 0;
+    let disabled = 0;
+
+    for (const p of providersFromCdn) {
+      allCdnIdentities.add(p.identity);
+      p.identityHistory.forEach((h) => allCdnIdentities.add(h));
+    }
+
+    for (const cdnProvider of providersFromCdn) {
+      const possibleIds = [cdnProvider.identity, ...cdnProvider.identityHistory];
+
+      const matchingCurrent =
+        possibleIds.map((id) => currentProvidersMap.get(id)).find(Boolean) ??
+        null;
+
+      if (matchingCurrent) {
+        const shouldUpdateIdentity =
+          matchingCurrent.identity !== cdnProvider.identity;
+        const shouldUpdateName = matchingCurrent.name !== cdnProvider.name;
+        const shouldUpdateUrl = matchingCurrent.url !== cdnProvider.url;
+
+        if (shouldUpdateIdentity || shouldUpdateName || shouldUpdateUrl) {
+          await db
+            .update(providersTable)
+            .set({
+              identity: cdnProvider.identity,
+              name: cdnProvider.name,
+              url: cdnProvider.url,
+              updatedBy: userIdentity,
+            })
+            .where(eq(providersTable.id, matchingCurrent.id));
+          updated += 1;
+        }
+      } else {
+        await db.insert(providersTable).values({
+          name: cdnProvider.name,
+          identity: cdnProvider.identity,
+          url: cdnProvider.url,
+          enabled: false,
+          visible: false,
+          createdBy: userIdentity,
+          updatedBy: userIdentity,
+        });
+        inserted += 1;
+      }
     }
-    return await response.json();
+
+    for (const provider of currentProviders) {
+      if (!allCdnIdentities.has(provider.identity) && (provider.enabled || provider.visible)) {
+        await db
+          .update(providersTable)
+          .set({
+            enabled: false,
+            visible: false,
+            updatedAt: new Date(),
+            updatedBy: userIdentity,
+          })
+          .where(eq(providersTable.identity, provider.identity));
+        disabled += 1;
+      }
+    }
+
+    console.log(
+      `[Providers] Done. Inserted: ${inserted}, Updated: ${updated}, Disabled: ${disabled}`,
+    );
```

Comment:
I know it can be tricky but maybe all this db operations must be made with a transaction. If some error happens, some providers will be updated while others not.